data/trie/patriciaMerkleTrie.go

```diff
@@ -445,37 +449,71 @@ func (tr *patriciaMerkleTrie) ResetOldHashes() [][]byte {
 	return oldHashes
 }
 
-// Snapshot creates a new database in which the current state of the trie is saved.
-// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
-func (tr *patriciaMerkleTrie) Snapshot() error {
+// Checkpoint adds the current state of the trie to the snapshot database
+func (tr *patriciaMerkleTrie) Checkpoint() error {
 	tr.mutOperation.Lock()
 	defer tr.mutOperation.Unlock()
 
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
 	if tr.snapshotInProgress {
 		return ErrSnapshotInProgress
 	}
 	tr.snapshotInProgress = true
 
-	err := tr.root.commit(false, 0, tr.db, tr.db)
+	if len(tr.snapshots) == 0 {
+		_, err := tr.newSnapshotDb()
+		if err != nil {
+			tr.snapshotInProgress = false
+			return err
+		}
+	}
+
+	lastSnapshotId := len(tr.snapshots) - 1
+	err := tr.root.commit(true, 0, tr.db, tr.snapshots[lastSnapshotId])
 	if err != nil {
+		tr.snapshotInProgress = false
 		return err
 	}
 
+	keys := tr.pruningBuffer
+	tr.pruningBuffer = make([][]byte, 0)
+	tr.snapshotInProgress = false
+
+	for i := range keys {
+		err = tr.removeFromDb(keys[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Snapshot creates a new database in which the current state of the trie is saved.
+// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
+func (tr *patriciaMerkleTrie) Snapshot() error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
+	if tr.snapshotInProgress {
+		return ErrSnapshotInProgress
+	}
+	tr.snapshotInProgress = true
+
 	rootHash := tr.root.getHash()
 
-	db, err := storageUnit.NewDB(
-		storageUnit.DBType(tr.snapshotDbCfg.Type),
-		path.Join(tr.snapshotDbCfg.FilePath, strconv.Itoa(tr.snapshotId)),
-		tr.snapshotDbCfg.BatchDelaySeconds,
-		tr.snapshotDbCfg.MaxBatchSize,
-		tr.snapshotDbCfg.MaxOpenFiles,
-	)
+	db, err := tr.newSnapshotDb()
 	if err != nil {
+		tr.snapshotInProgress = false
```

Comment:
can you instead defer the setting of snapshot in progress to false ? and remove this from all returns?

Reply:
+1. Also in Checkpoint method

Reply:
I can not use defer, because of the goroutine on line 554. I don't want to set snapshotInProgress to false when this function has finished execution, but when the snapshot() function that runs on a goroutine on line 554 will be done.

Reply:
In the checkpoint method I can and will use defer.

Reply:
Used defer on the Checkpoint method, but I left the `tr.snapshotInProgress = false` from L484, because the snapshoting process is done before executing L486-L491.